### PR TITLE
Fix empty proposal errors in the training of some two-stage models

### PR DESCRIPTION
**Pull request by hhaAndroid**
## Motivation

Fix empty proposal errors in the training of some two-stage models.

Resolve: https://github.com/open-mmlab/mmdetection/issues/5917

### CODE REVIEW

File: mmdet/models/roi_heads/scnet_roi_head.py [hunk 1]
```diff
@@ -102,7 +102,9 @@ def _bbox_forward(self,
         bbox_head = self.bbox_head[stage]
         bbox_feats = bbox_roi_extractor(
             x[:len(bbox_roi_extractor.featmap_strides)], rois)
-        if self.with_semantic and semantic_feat is not None:
+        # bbox_feats.shape[0] > 0 is mean the number of proposal is not 0.
```
[the remark this answers is not in the document]
Comment (jshilong) [reply mmdet/models/roi_heads/scnet_roi_head.py:105]: This would raise an error in DDP because there are parameters not used in forward

**Comment by jshilong** [comment]
Please make sure all parameters should be used in forwarding process

**Comment by hhaAndroid** [comment]
@jshilong Please help deal with the problem of the sparse-rcnn empty proposal.

**Comment by ZwwWayne** [comment]
> @jshilong Please help deal with the problem of the sparse-rcnn empty proposal.

sparse rcnn should not have such a problem because it always has a constant number of proposals？

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 2]
```diff
@@ -108,8 +108,12 @@ def custom_accuracy(self):
     @auto_fp16()
     def forward(self, x):
         if self.with_avg_pool:
-            x = self.avg_pool(x)
-        x = x.view(x.size(0), -1)
+            if x.shape[0] > 0:
+                x = self.avg_pool(x)
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:112]: 1. May indicate why we need to do that, e.g., avg_pool does not support empty tensor 
2. may use x.numel() rather than x.shape[0] for consistency

File: mmdet/models/roi_heads/cascade_roi_head.py [hunk 3]
```diff
@@ -271,6 +271,11 @@ def forward_train(self,
                     if self.bbox_head[i].custom_activation:
                         cls_score = self.bbox_head[i].loss_cls.get_activation(
                             cls_score)
+
+                    # Empty proposal.
+                    if cls_score.shape[0] == 0:
+                        break
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/cascade_roi_head.py:277]: Simply break here might cause different graph during training?

File: mmdet/models/utils/brick_wrappers.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn.bricks.wrappers import NewEmptyTensorOp, obsolete_torch_version
+
+if torch.__version__ == 'parrots':
+    TORCH_VERSION = torch.__version__
+else:
+    # torch.__version__ could be 1.3.1+cu92, we only need the first two
+    # for comparison
+    TORCH_VERSION = tuple(int(x) for x in torch.__version__.split('.')[:2])
```
Comment (jshilong) [inline mmdet/models/utils/brick_wrappers.py:11]: You can use 
https://github.com/open-mmlab/mmdetection/blob/b1f97c1b6d2b3c735d8c5690952264c598c2f206/mmdet/__init__.py#L7

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 5]
```diff
@@ -108,8 +108,12 @@ def custom_accuracy(self):
...
+                x = x.view(x.size(0), -1)
+            else:
+                # Empty proposal
+                x = torch.mean(x, dim=(-1, -2))
```
Comment (jshilong) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:116]: Why here does not be tested

File: mmdet/models/roi_heads/cascade_roi_head.py (hunk 3 above)
Comment (hhaAndroid) [reply mmdet/models/roi_heads/cascade_roi_head.py:277]: This is the refining stage, no gradient is needed. It should not cause deadlock.

File: mmdet/models/utils/brick_wrappers.py (hunk 4 above)
Comment (hhaAndroid) [reply mmdet/models/utils/brick_wrappers.py:11]: We just follow https://github.com/open-mmlab/mmcv/blob/master/mmcv/cnn/bricks/wrappers.py#L16.  If digit_version is adopted, I need special treatment for parrots.

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py (hunk 5 above)
Comment (hhaAndroid) [reply mmdet/models/roi_heads/bbox_heads/bbox_head.py:116]: Because `self.with_avg_pool` is only set to True for certain algorithms, unit tests are not added.

**Review by ZwwWayne — APPROVED** [review]
